packages/address-book-controller/src/AddressBookController.ts

```diff
@@ -153,6 +195,184 @@ export class AddressBookController extends BaseController<
 
     return true;
   }
+
+  /**
+   * Initialization step which optionally listens to changes to frequent RPC
+   * chain IDs in {@link PreferencesController} state and ensures that
+   * addresses are filed under the correct chain.
+   */
+  #handleChainIdChanges() {
+    const { syncWithRpcChanges, onPreferencesStateChange, getProvider } =
+      this.config;
+
+    if (syncWithRpcChanges) {
+      if (onPreferencesStateChange === undefined) {
+        throw new Error(
+          'onPreferencesStateChange is required along with syncWithRpcChanges.',
+        );
+      }
+
+      if (getProvider === undefined) {
+        throw new Error(
+          'getProvider is required along with syncWithRpcChanges.',
+        );
+      }
+
+      onPreferencesStateChange(
+        (newPreferencesState, previousPreferencesState) => {
+          return this.#handlePreferencesStateChange(
+            previousPreferencesState,
+            newPreferencesState,
+            getProvider,
+          ).catch((error) => {
+            // TODO: Should this use log?
+            console.error(error);
+          });
+        },
+      );
+    }
+  }
+
+  /**
+   * Callback to `onPreferencesStateChange` which detects changes to frequent
+   * RPC chain IDs within {@link PreferencesController} state and migrates
```

Comment:
I don't think we need to migrate addresses anymore. This was meant to accommodate a specific update for metamask-extension, where we went from `chainId` being optional to being required, and where we switched from keying these entries by `chainId` instead of `networkId`.

That was over two years ago. The only wallets that old will be dormant ones. For the tiny subset of users that are starting up a wallet that has been shut down for multiple years, I think we can consider the _possibility_ of losing addresses to be an acceptable risk. Even at the time of the migration this was an edge case.

Reply:
Instead we could watch for list changes and _delete_ address book entries that are keyed by a chainID no longer being used.

Or, even easier, we could do nothing at let the addresses become orphaned when a chain is deleted. Then they can be seen again if the chain is re-added.